lib/Doctrine/Common/Cache/CouchbaseBucketCache.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\Common\Cache;
+
+use Couchbase\Bucket;
+use Couchbase\Document;
+use Couchbase\Exception;
+
+/**
+ * Couchbase ^2.3.0 cache provider.
+ */
+class CouchbaseBucketCache extends CacheProvider
+{
+    /**
+     * @var Bucket
+     */
+    private $bucket;
+
+    /**
+     * CouchbaseCache constructor.
+     * @param Bucket $bucket
+     */
+    public function __construct(Bucket $bucket)
+    {
+        $this->bucket = $bucket;
+
+        if (!is_callable([$bucket, 'manager'])) {
+            // Manager is required to flush cache and pull stats.
+            throw new \RuntimeException('ext-couchbase:^2.3.0 is required.');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doFetch($id)
+    {
+        $id = $this->normalizeKey($id);
+
+        try {
+            $document = $this->bucket->get($id);
+        } catch (Exception $e) {
+            // TODO Log exception.
+            return false;
+        }
+
+        if ($document instanceof Document && $document->value !== false) {
+            return $this->decode($document->value);
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doContains($id)
+    {
+        $id = $this->normalizeKey($id);
+
+        try {
+            $document = $this->bucket->get($id);
+        } catch (Exception $e) {
+            // TODO Log exception.
+            return false;
+        }
+
+        if ($document instanceof Document) {
+            return !$document->error;
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doSave($id, $data, $lifeTime = 0)
+    {
+        $id = $this->normalizeKey($id);
+
+        if ($lifeTime > 30 * 24 * 3600) {
```

Comment:
originally, i just copied from [\Doctrine\Common\Cache\CouchbaseCache](https://github.com/doctrine/cache/blob/fe68e4f78804684e31cce17590b0bd9f088db1b0/lib/Doctrine/Common/Cache/CouchbaseCache.php#L81), but then i saw in the PHPStorm `\Couchbase\Bucket::upsert()` stub that expiry > 30 days is treated in a special way.

here are the real [Couchbase expiry docs](https://developer.couchbase.com/documentation/server/4.1/developer-guide/expiry.html), which say:
```
If the absolute value of the expiry is less than 30 days (i.e. 60 * 60 * 24 * 30), it is considered an offset. If it is greater, it is considered an absolute time stamp.
```

Couchbase Server uses Memcached internally, and exposes a Memcached API to connect to specially configured buckets via their Moxi service. i think that the strange expiry behavior in Couchbase is meant to maintain compatibility with the Memcached API.

you can see the same special handling of expiry in the PHP Memcached docs and in the Memcached cache providers in this library:
* http://php.net/manual/en/memcache.set.php
* http://php.net/manual/en/memcached.expiration.php
* https://github.com/doctrine/cache/blob/206b6346465133b5939ffcf255d26559679a1da7/lib/Doctrine/Common/Cache/MemcachedCache.php#L107
* https://github.com/doctrine/cache/blob/fe68e4f78804684e31cce17590b0bd9f088db1b0/lib/Doctrine/Common/Cache/MemcacheCache.php#L89